grails-app/services/org/pih/warehouse/inventory/ProductAvailabilityService.groovy

```diff
@@ -776,80 +782,84 @@ class ProductAvailabilityService {
+                if (!command.showOutOfStockProducts) {
+                    // Read: 0 is less than result from subquery
+                    add Subqueries.lt(0, aggregatedQuantityQuery)
                 }
-                // This is pretty inefficient if the previous query does not narrow the results
-                // if the inner products list is empty, but there are search terms then return empty results
-                if (innerProductIds || searchTerms) {
-                    'in'("id", innerProductIds ?: [null])
+            }
+            // Or apply default product type restrictions:
+            // return products with product type with activity searchable no stock OR (searchable AND qoh > 0)
+            else {
+                if (!command.showOutOfStockProducts) {
+                    add(Restrictions.disjunction().
+                            add(Restrictions.in("productType", searchableNoStockProductTypes)).
+                            add(Restrictions.conjunction().
+                                    add(Subqueries.lt(0, aggregatedQuantityQuery)).
+                                    add(Restrictions.in("productType", searchableProductTypes))
+                            ))
                 }
             }
-        }
 
-        def quantityMap = products ? getQuantityOnHandByProduct(command.location, products) : []
-
-        def items = []
+            // Restrict products by selected categories
+            if (categories) {
+                'in'("category", categories)
+            }
 
-        products.each { Product product ->
-            def quantity = quantityMap[product] ?: 0
+            // Restrict products by selected tags
+            if (command.tags) {
+                tags {
+                    'in'("id", command.tags*.id)
+                }
+            }
 
-            if (product.productType && !product.productType.supportedActivities?.contains(ProductActivityCode.SEARCHABLE_NO_STOCK)) {
-                if (!product.productType.supportedActivities?.contains(ProductActivityCode.SEARCHABLE)) {
-                    return
-                } else if (quantity == 0) {
-                    return
+            // Restrict products by selected catalogs
+            if (command.catalogs) {
+                productCatalogItems {
+                    productCatalog {
+                        'in'("id", command.catalogs*.id)
+                    }
                 }
             }
 
-            items << [
+            // This is pretty inefficient if the previous query does not narrow the results
+            // if the inner products list is empty, but there are search terms then return empty results
+            if (innerProductIds || searchTerms) {
+                'in'("id", innerProductIds ?: [null])
+            }
+            order("productCode")
+        }
+
+        def quantityMap = products ? getQuantityOnHandByProduct(command.location, products) : [:]
```

Comment:
if I understand the query correctly - I assume that in the `createCriteria` we already count the `quantityOnHandByProduct` - so do we have to do it also here at this point?